akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -53,20 +55,31 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
   implicit def system: ActorSystem[_]
   implicit def executionContext: ExecutionContext
 
+  var telemetry: Telemetry = _
+
   def readOffsets(): Future[Option[Offset]]
   def saveOffset(projectionId: ProjectionId, offset: Offset): Future[Done]
   val killSwitch: SharedKillSwitch = KillSwitches.shared(projectionId.id)
   val abort: Promise[Done] = Promise()
 
-  private def reportProgress[T](after: Future[T], env: Envelope): Future[T] = {
-    after.map { done =>
-      try {
-        statusObserver.progress(projectionId, env)
-      } catch {
-        case NonFatal(_) => // ignore
+  private def reportProgress[T](after: Future[T], envelope: Envelope): Future[T] = {
+    after
+      .map { done =>
+        try {
+          statusObserver.progress(projectionId, envelope)
+        } catch {
+          case NonFatal(_) => // ignore
+        }
+        done
+      }
+  }
+
+  private def saveOffsetWithMetrics(projectionId: ProjectionId, offset: Offset, batchSize: Int): Future[Done] = {
```

Comment:
I'm not sure we can do that. `reportProgress` takes an `after: Future[T]` which, sometimes is only a `saveOffset` and sometimes is a `saveOffset().flatMap(handler.process(envelope))`. 

Note, `saveOffset` and `reportProgress` were already separate before this PR and I just wrapped `saveOffset` with the telemetry invocation.

The difference, I think, is what `StatusObserver#progress` and `Telemetry#onOffsetStored` mean:

```
  /**
   * Called when the envelope has been processed and corresponding offset stored.
   * It might not be called for each envelope.
   */
  def progress(projectionId: ProjectionId, env: Envelope): Unit
```
So it is safe to always invoke `StatusObserver#progress` last.

Instead,

```
  /**
   * Invoked when the offset is committed.
   */
  def onOffsetStored(numberOfEnvelopes: Int): Unit
```

so Telemetry needs tp bind that invocation to the offset being stored (even if the envelope processing didn't happen yet).

Reply:
The difference is in exactly-once, right? we can change the semantics of `progress` to be the same as `onOffsetStored`. I think we discussed that renaming in some other ticket already.

Reply:
I was thinking about `at-most-once` where the `after` in `private def reportProgress[T](after: Future[T], envelope: Envelope)` is all of the following: (1) saving the offset, (2) processing the envelope/s and (3) retry (aka recovery handling). When batching, if there's an error in some envelope telemetry should still report one offset was stored and that _so many_ envelopes were processed. If, for example, the recovery strategy is `fail` then telemetry will still report the offset stored, but `statusObserver.progress` in `reportProgress` will report nothing.

This is one of the current differences between `StatusObserver` and `Telemetry`. 

 - `StatusObserver` only provides `progress` which includes the last `Envelope` processed. In case of an error when processing a group or a batch it may not even be invoked.
 - `Telemetry ` provides `afterProcess` and `onOffsetStored` which are invoked as many times as envelopes processed of offset store operations (respectively). It's a heavier approach but chances for missed invocations are smaller.

Reply:
ok, we can change the semantics of `StatusObserver.progress` to be the same as `Telemetry.onOffsetStored`.